### ISSUE TITLE
Did you align faces when computing ID loss?

### ISSUE THREAD

**Issue by 18445864529** [OrLichter/lcm-lookahead#3]
i noticed this variable in the IDLoss code `self.reference_facial_points`
but it was nowhere used. And I didn't find the code for aligning faces before extracting feats and computing id losses.
Please point it out if I'm missing something. Thanks.

**Comment by rinongal** [comment]
It's happening in the extract_faces_and_landmarks function [here](https://github.com/OrLichter/lcm-lookahead/blob/fc0154686adee4458cf05cc669fa1ef80d8d79cc/utils/utils.py#L40C5-L40C32) which is called when computing the ID loss [here](https://github.com/OrLichter/lcm-lookahead/blob/fc0154686adee4458cf05cc669fa1ef80d8d79cc/losses/losses.py#L210).

Seems like the self.reference_facial_points line is just duplicated unused code (you'll see the above linked file has the the same reference points defined and used).

**Comment by 18445864529** [comment]
@rinongal Thanks! Your reply is faster than diffusion models! I see now.
Just to make sure, the alignment operation performed here can keep the gradient flow right? 
(the `from kornia.geometry.transform.crop2d import warp_affine`)

**Comment by rinongal** [comment]
Yep. Gradients get through, and you'll see that when we call it in the loss function we wrap it in no_grad for the encoder's input image. Another important bit is that the alignment often fails, so we also return an indicator of which batch entries failed it and we don't compute the identity loss for those.

**Comment by 18445864529** [comment]
awesome. thanks.

**Comment by 18445864529** [comment]
> Yep. Gradients get through, and you'll see that when we call it in the loss function we wrap it in no_grad for the encoder's input image.

I'm curious whether we can put everything in the ID loss computation under torch.no_grad() since these models are just for inference purposes and not trained. A more general question would be, if I train only my UNet but need to VAE decode its output and then pass through some face modules to lead to the final loss, and if I put all these operations under torch.no_grad(), will it affect the training result at all?

**Comment by rinongal** [comment]
You cannot put everything under no_grad since we use the identity network as a loss for training the encoder. 
We have to backpropagate through it.

**Comment by 18445864529** [comment]
but even putting vae and face modules under torch.no_grad can still allow the backprop to be performed, since all your trainable params are in the unet part, is it not?

i was originally thinking and doing the same way as you do but copilot told me the other way. she said that if we do `unet -> vae -> face modules -> id loss` and only unet is updated, it is ok to put vae and face modules under no_grad, because it only disables the grad calculations for these modules but not prevent the gradient flow from the id loss to the unet. 

So I tried (need to manually call `id_loss.requires_grad_(True)` before backward), and it worked.

I do not fully trust her but it just worked. That's why I want to make sure about this with you whether this could introduce any incorrectness. Because after all, if not, it can save up a lot of memory.

**Comment by rinongal** [comment]
Can you verify that you're actually getting gradients to the encoder? As far as I am aware, no_grad used to block gradients to earlier parts of the network (see e.g., examples [here](https://stackoverflow.com/questions/51748138/pytorch-how-to-set-requires-grad-false) or cell 9 in [this notebook](https://github.com/MeiqiGuo/PytorchLearningNotes/blob/master/NoGrad.ipynb)) but admittedly I did not try it myself.

**Comment by 18445864529** [comment]
I think the link you provided is more likely to be the truth. That's why there's a cost to trusting LLMs. Maybe even there was no error for that implementation, the grads for the ID branch were actually not propagated through the UNet but only the SD loss was taking effect. Thanks for sharing this post! I will do more exps and close this if everything's going well.
